src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs

```diff
@@ -210,36 +212,81 @@ private List<ProviderSpecificWorkload<TestRunCriteria>> SplitToWorkloads(TestRun
         // is shared we try to find a workload that is appropriate for it. If we don't find any work that the
         // running testhost can do. Or if the testhost already exited (possibly because of crash), we start another one
         // and give it the next workload.
+
+
+        // List all the sources (dlls) we have and group them by their provider, so we can run either multiple sources
+        // on a single instance of the appropriate testhost (for non-paralell, shared),
+        // or each source in its own instance of the provider (for non-parallel, non-shared, and both parallel shared, and parallel non-shared).
         List<ProviderSpecificWorkload<TestRunCriteria>> workloads = new();
         if (testRunCriteria.HasSpecificTests)
         {
             // We split test cases to their respective sources, and associate them with additional info about on
             // which type of provider they can run so we can later select the correct workload for the provider
             // if we already have a shared provider running, that can take more sources.
-            var testCasesPerSource = testRunCriteria.Tests.GroupBy(t => t.Source);
-            foreach (var group in testCasesPerSource)
+            Dictionary<string, List<TestCase>> sourceToTestCasesMap = testRunCriteria.Tests.GroupBy(t => t.Source).ToDictionary(k => k.Key, v => v.ToList());
+
+            var sources = sourceToTestCasesMap.Keys;
+            // Each source is grouped with its respective provider.
+            var providerGroups = sources.Select(source => new ProviderSpecificWorkload<string>(source, sourceToTestHostProviderMap[source])).GroupBy(psw => psw.Provider);
+
+            foreach (var group in providerGroups)
             {
-                var testHostProviderInfo = sourceToTestHostProviderMap[group.Key];
-                var runsettings = testHostProviderInfo.RunSettings;
-                // ToList because it is easier to see what is going on when debugging.
-                var testCases = group.ToList();
-                var updatedCriteria = CreateTestRunCriteriaFromTestCasesAndSettings(testCases, testRunCriteria, runsettings);
-                var workload = new ProviderSpecificWorkload<TestRunCriteria>(updatedCriteria, testHostProviderInfo);
-                workloads.Add(workload);
+                var testhostProviderInfo = group.Key;
+                // If the run is not parallel and the host is shared, put all testcases on single testhost.
```

Comment:
Most of my remarks for discovery can also be applied here.

Reply:
also applied here.